lib/Driver/ToolChains.cpp

```diff
@@ -1340,6 +1360,42 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
   return {"clang++", Arguments};
 }
 
+std::string
+toolchains::Android::getTarget() const {
+  // Explicitly set the linker target to "androideabi", as opposed to the
+  // llvm::Triple representation of "armv7-none-linux-android".
+  // This is the only ABI we currently support for Android.
+  return "armv7-none-linux-androideabi";
+}
+
+bool toolchains::Android::shouldSpecifyTargetTripleToLinker() const {
+  return false;
+}
+
+bool toolchains::Android::shouldProvideRPathToLinker() const {
+  return false;
+}
+
+std::vector<std::string> toolchains::Android::getAdditionalLinkerPaths() const {
+  // "ANDROID_NDK_HOME" is set by the Swift build script. Still, emit a nice
+  // assertion message here for people building via CMake directly (even though
+  // this is not something we support).
+  const char *ndkhome = getenv("ANDROID_NDK_HOME");
+  assert(ndkhome && "Environment variable ANDROID_NDK_HOME needs to be set "
+    "when building Swift to target Android. It should point to the directory "
+    "where the Android NDK has been installed, such as "
+    "/usr/local/opt/android-ndk-r10e.");
+
+  // FIXME: This ties the Android driver to the tree layout of Android's NDK,
+  //        which is undesirable. It may be better to allow the build script
+  //        to specify a path to these instead. This would also allow a
+  //        different libc++ from within the Android NDK to be specified. 
+  auto libgcc = Twine(ndkhome) + "/toolchains/arm-linux-androideabi-4.8/"
```

Comment:
Twines are temporary objects, unfortunately; they're not guaranteed to work correctly if stored in local variables. You'll have to do this all in one expression.